app/controllers/api/v1/base_controller.rb

```diff
@@ -8,20 +8,34 @@ class BaseController < ApplicationController
       private
 
       def require_authentication!
-        user = User.find(request.headers["X-User-Id"])
+        authenticator = Authenticator.new(
+          user: User.find_by(id: request.headers["X-User-Id"]),
+          provided_token: request.headers["Authorization"]
+        )
 
-        if user.present? && valid_token?(user.authentication_token)
-          sign_in user
+        if authenticator.valid?
+          sign_in authenticator.user
         else
-          render status: :unauthorized
+          head :unauthorized
         end
       end
 
-      def valid_token?(token)
-        ActiveSupport::SecurityUtils.secure_compare(
-          "Token #{token}",
-          request.headers["Authorization"]
-        )
+      class Authenticator
+        attr_reader :user, :provided_token
+
+        def initialize(user:, provided_token:)
```

Comment:
would renaming `provided_token` to `token` make this an ever so slightly better name 🤔 
The attribute being prefixed with `provided` makes sense to indicate where it comes from, but it bothers me a little that the name of the parameter has the prefix 🤔

Reply:
I always want to be explicit about the token that we've been provided by the person trying to log in and any token that might be valid for the user. Just calling it `token` introduces some ambiguity.
If you see it in isolation:

`Authenticator.new(user: whoever, token: '123xyz')`

Is this going to be used for comparing? Is it assigning a token to be used for that user which will be validated later some other way?

`provided_token` name still bugs me a bit because it's still a bit unclear what "provided" might mean.

All this is probably a sign that there's some refactoring hiding in here that could make things clearer.

Reply:
Can similar questions be raised about `user`?
To me it appears relatively clear that the `Authenticator`  will try to authenticate the `user` with said `token` 🤔 
I see your argument though... the `Authenticator` could assign the `token` to the `user` if the `user` is valid, name wise.

sneaky one 👍 
thanks ^__^

Reply:
> All this is probably a sign that there's some refactoring hiding in here that could make things clearer.

See the next 2 commits where I do this refactoring.